wandb/apis/public.py

```diff
@@ -2136,6 +2087,183 @@ def __repr__(self):
+        for item in res["project"]["runQueue"]["runQueueItems"]["edges"]:
+            if str(item["node"]["id"]) == str(self.id):
+                return item["node"]["state"].lower()
+        raise ValueError(
+            f"Could not find QueuedRun associated with id: {self.id} on queue id {self.queue_id}"
+        )
+
+    def _exec(self, query, **kwargs):
+        """Execute a query against the cloud backend"""
+        variables = {
+            "entity": self.entity,
+            "project": self.project,
+        }
+        variables.update(kwargs)
+        return self.client.execute(query, variable_values=variables)
+
+    @normalize_exceptions
+    def wait_until_finished(self):
+        if not self._run:
+            self.wait_until_running()
+
+        self._run.wait_until_finished()
+        # refetch run to get updated summary
+        self._run.load(force=True)
+        return self._run
+
+    @normalize_exceptions
+    def delete(self, delete_artifacts=False):
+        """
+        Deletes the given queued run from the wandb backend.
+        """
+        mutation = gql(
+            """
+            mutation DeleteFromRunQueue($queueID: String!, $runQueueItemID: String!)
+            {
+                deleteFromRunQueue(input: {queueID: $queueID, runQueueItemID: $runQueueItemID}) {
+                    success
+                    clientMutationId
+                }
+            }
+            """
+        )
+        self.client.execute(
+            mutation,
+            variable_values={
+                "queueID": self.queue_id,
+                "runQueueItemID": self._run_queue_item_id,
+            },
+        )
+
+    @normalize_exceptions
+    def wait_until_running(self):
+        if self._run is not None:
+            return self._run
+        if self.container_job:
+            raise LaunchError("Container jobs cannot be waited on")
+        query = gql(
+            """
+            query GetRunQueueItem($projectName: String!, $entityName: String!, $runQueue: String!) {
+                project(name: $projectName, entityName: $entityName) {
+                    runQueue(name:$runQueue) {
+                        runQueueItems {
+                            edges {
+                                node {
+                                    id
+                                    state
+                                    associatedRunId
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+        """
+        )
+        variable_values = {
+            "projectName": self.project,
+            "entityName": self._entity,
+            "runQueue": self.queue_id,
+        }
+
+        while True:
+            res = self.client.execute(query, variable_values)
+            # TODO: add fetch run queue by item end point
+            for item in res["project"]["runQueue"]["runQueueItems"]["edges"]:
```

Comment:
In the off-chance that the original query does not return a matching node, then i think we hit an inf loop? I think if the inner if block is not entered on the first iteration, we should error/break early